Sources/GRPC/Server.swift

```diff
@@ -490,10 +527,25 @@ extension Server.Configuration.CORS {
     /// Allow all origin values.
     public static let all = Self(.all)
 
+    /// Allow all origin values; similar to `all` but returns the value of the origin header field
+    /// in the 'access-control-allow-origin' response header (rather than "*").
+    public static let originBased = Self(.originBased)
+
     /// Allow only the given origin values.
     public static func only(_ allowed: [String]) -> Self {
       return Self(.only(allowed))
     }
+
+    /// Provide a custom CORS origin check.
+    ///
+    /// - Parameter checkOrigin: A closure which is called with the value of the 'origin' header
+    ///     and returns the value to use in the 'access-control-allow-origin' response header,
+    ///     or `nil` if the origin is not allowed.
+    public static func custom(
+      _ checkOrigin: @escaping @Sendable (_ origin: String) -> String?
+    ) -> Self {
```

Comment:
Given the requirement for sendability and hashability, should this take a generic type or something equivalent to enable you to maintain the invariant?

Reply:
Hmm, we'd need to store it as an existential which means we'd get the automatic `Hashable` conformance but not the `Equatable` conformance (the stored `any P`s could be of different types). It's also mildly inconvenient for callers who are implementing that type. My lean is towards not making this generic.

Reply:
We can get the equatable and hashable conformances using `AnyHashable` and more generally can achieve all of this using closures. The question is whether it's a good idea to keep types that don't really behave the way we want them to, I think.